packages/e2e-tests/helpers/elements-panel.ts

```diff
@@ -9,78 +9,131 @@ export async function activateInspectorTool(screen: Screen): Promise<void> {
-            const overridden = (selectorNode.parentNode as HTMLElement).classList.contains(
-              "computed-overridden"
-            );
-            return { selector, value, label, url, overridden };
-          });
-        }
+export async function expandPseudoElementRules(screen: Screen) {
+  const header = screen.locator("#rules-section-pseudoelement-header");
+  if ((await header.count()) === 0) {
+    return;
+  }
+  const classes = await header.locator(".theme-twisty").getAttribute("class");
+  if (!classes?.includes("open")) {
+    await header.click();
+  }
+}
 
-        const names = document.querySelectorAll<HTMLElement>(".computed-property-name");
-        const propertyName = [...names].find(n => n!.textContent!.includes(style));
-        if (!propertyName) {
-          return false;
-        }
-        const container = propertyName.closest(".computed-property-view");
-        if (!container) {
-          return false;
+export async function getAppliedRules(screen: Screen) {
+  await openAppliedRulesTab(screen);
+  await expandPseudoElementRules(screen);
+  return await screen.evaluate(() => {
+    const rules = document.querySelectorAll<HTMLElement>(".ruleview-rule");
+    return [...rules].map(rule => {
+      const selector = rule
+        .querySelector<HTMLElement>(".ruleview-selectorcontainer")!
+        .innerText.trim();
+      const source = rule.querySelector<HTMLElement>(".ruleview-rule-source")!.innerText.trim();
+      const properties = [...rule.querySelectorAll<HTMLElement>(".ruleview-propertycontainer")].map(
+        prop => {
+          let longhandProps;
+          if (prop.nextSibling) {
+            longhandProps = [
+              ...(prop.nextSibling as HTMLElement).querySelectorAll<HTMLElement>("li"),
+            ].map(longhand => ({
+              text: longhand.innerText,
+              overridden: longhand.classList.contains("ruleview-overridden"),
+            }));
+          }
+          const result: any = {
+            text: prop.innerText.trim(),
+            overridden: (prop.parentNode as HTMLElement).className.includes("overridden"),
+          };
+          if (longhandProps) {
+            result.longhandProps = longhandProps;
+          }
+          return result;
         }
-        const propertyValue: HTMLElement | null = container.querySelector(
-          ".computed-property-value"
-        );
-        const expectedSelectorsJSON = matchedSelectors
-          ? JSON.stringify(matchedSelectors)
-          : undefined;
-        const selectorsJSON = matchedSelectors
-          ? JSON.stringify(getMatchedSelectors(style))
-          : undefined;
-        return (
-          propertyValue!.textContent!.includes(value) && selectorsJSON === expectedSelectorsJSON
-        );
-      },
-      { style, value, matchedSelectors: expectedSelectors }
-    );
-    expect(result).toBe(true);
+      );
+      return { selector, source, properties };
+    });
   });
 }
```

Comment:
These functions are all pretty scary... sibling nodes, parent nodes– all traversing the structure the DOM _happens to be_.

I wonder if we could write the helper functions without being so tightly coupled to the DOM structure?